mahiru/components/domain_administrator.py

```diff
@@ -180,6 +186,88 @@ def execute_step(
                 for asset in assets.values():
                     self._free_image(asset.id)
 
+    def serve_asset(
+            self, asset: Asset, connection_request: ConnectionRequest
+            ) -> ConnectionInfo:
+        """Serve an asset as a VPN-reachable service.
+
+        The asset's image_location must be a path on the local disk,
+        this function does not serve remote images.
+
+        Args:
+            asset: A local asset to serve.
+            connection_request: A description of the remote end of the
+                VPN connection to set up.
+        """
+        if asset.image_location is None:
+            raise RuntimeError(
+                    f'Cannot serve asset {asset} because it has no image')
+
+        if asset.image_location.startswith('http:'):
+            raise RuntimeError(f'Refusing to serve remote asset {asset}')
+
+        job_id = self._get_job_id()
+        conn_id = str(job_id)
+
+        logger.info(f'Serving container {asset.id} as job {job_id}')
+
+        image = None
+        container = None
+        try:
+            image = self._ensure_image_available(asset)
+
+            container_name = f'mahiru-{job_id}-data-asset-remote'
+            container = self._dcli.containers.run(
+                    image, name=container_name,
+                    detach=True, network_mode='none')
+            container.reload()
+            pid = container.attrs['State']['Pid']
+
+            with self._served_lock:
+                self._served_assets[conn_id] = asset.id
+                self._served_containers[conn_id] = container
+
+            return self._network_administrator.serve_asset(
+                    conn_id, pid, connection_request)
+
+        except Exception as e:
+            logger.error(f'Error serving asset connection: {e}')
+            if container is not None:
+                self._remove_containers([container])
+            if image is not None:
+                self._free_image(asset.id)
+            if conn_id in self._served_assets:
+                del self._served_assets[conn_id]
+            if conn_id in self._served_containers:
+                del self._served_containers[conn_id]
+            raise
+
+    def stop_serving_asset(self, conn_id: str) -> None:
+        """Stop and remove an asset container being served remotely.
+
+        The conn_id argument must have been returned by serve_asset()
+        previously. Implementation note: it's just the job id converted
```

Comment:
Okay, this is my fourth attempt at writing an explanation of what's going on. Let's see if it goes better. When our site executes a step (a job that we do), we need to connect to each input. For each input, this creates a connection between the sites, and a job serving the asset at the remote end. So if we have a step with two inputs both served remotely from different sites, then we have three jobs running at three different sites, and two connections.

From the perspective of the site running the step, we have a local job for the step with a local job id. That is an internal id we use to keep track of our local jobs and give their containers unique names. When we connect to the input assets, we ask the remote site to create a connection for us. It does that, gives it a connection id, and then passes that connection id back to us, so that we can later use it to identify the connection when we want them to shut it down again.

Now, internally, they consider serving an asset to be a job, for which they generate a job id so they can keep the container apart from the ones used for other (step execution or asset serving) jobs. But we don't know that, and anyway we don't consider that job to the the one associated with the connection. For us, that is our local step running job!

And so that's why the two are not the same thing. Now it so happens that the serving site, on shutting down the connection, needs to shut down the data asset container, for which it needs its name, which contains the job id. Job ids are unique within the site, so I can use it as the connection id directly and save myself a translation table from connection ids to job ids or container names. Of course, I do still need to track who requested which connection to make sure people can't shut down each other's connections, but that's a separate thing.